Set an upper limit of how large we will let txn-queues grow.

jameinel opened the pull request:
When we have broken transaction data in the database (such as from
mongo getting OOM killed), it can cause cascade failure, where that
document ends up getting too many transactions queued up against it.

This can also happen if you have nothing but assert-only transactions against a
single document.

If we have lots of transactions, it becomes harder and harder to add
new entries and clearing out a large queue is O(N^2) which means capping it
is worthwhile. (It also makes the document grow until it hits max-doc-size.)

The upper bound is still quite large, so it should not be triggered if
everything is operating normally.

jameinel committed:
Set an upper limit of how large we will let txn-queues grow.

When we have broken transaction data in the database (such as from
mongo getting OOM killed), it can cause cascade failure, where that
document ends up getting too many transactions queued up against it.

This can also happen if you have nothing but assert-only transactions against a
single document.

If we have lots of transactions, it becomes harder and harder to add
new entries and clearing out a large queue is O(N^2) which means capping it
is worthwhile. (It also makes the document grow until it hits max-doc-size.)

The upper bound is still quite large, so it should not be triggered if
everything is operating normally.

mjs commented on txn/flusher.go:
```diff
@@ -244,6 +246,16 @@ NextDoc:
 		chaos("")
 		if _, err := cquery.Apply(change, &info); err == nil {
+			if len(info.Queue) > maxTxnQueueLength {
+				// abort with TXN Queue too long, but remove the entry we just added
```
s/TXN Queue/txn-queue/ ?

mjs commented on txn/flusher.go:
```diff
@@ -212,6 +212,8 @@ var txnFields = bson.D{{"txn-queue", 1}, {"txn-revno", 1}, {"txn-remove", 1}, {"
 
 var errPreReqs = fmt.Errorf("transaction has pre-requisites and force is false")
 
+const maxTxnQueueLength = 1000
```
This seems like a sensible default but it might be good to add support for overriding it on a per session basis in case it's required for some application.

jameinel committed:
Add Runner.SetOptions to control maximum queue length.

Still defaults to 1000 without any other configuration, but
allows callers to know that they can be stricter/less strict.

fmpwizard commented:
If you close and reopen the PR, travis will rerun this build and thanks to  https://github.com/go-mgo/mgo/pull/462 , your PR should pass all tests on all mongodb versions

rogpeppe commented:
I don't understand this fully, but my immediate suspicion is that the code looks racy. I'd at least expect a doc comment explaining how/why it's not.

jameinel committed:
Switch from an error to aborting the txn.

jameinel committed:
Add a test for behavior when multiple documents participate with run that runs out of queue space.

jameinel committed:
Change the flusher to still return a custom error.

This still aborts the transaction, but then returns a custom error
instead of ErrAborted. That allows us to still log the details, which
lets us know what document is broken, but also avoids having a huge
queue of Pending transactions wanting to be applied.
